### PR TITLE
[BREAKING] fix: do not remove newly added blocks from main in swap_main_blockchain 

### PR DESCRIPTION
**Pull request by sander2**
The way `swap_main_blockchain` worked, was by traversing from the fork tip to the main chain by following the `prev_hash`es. It would move the blocks one by one to the main chain. When it reached the main chain, it moved all blocks that had a height higher than the first reached main block into the original fork. The problem is, it would move also the newly added blocks. As a result, blocks from the fork that overtakes main were unable to be used by transactions. 

With this PR, when a block is transferred to the main chain, the existing block is moved to into the fork. I also added some new invariants for `swap_main_blockchain`, basically checking that the blocks are moved correctly and agree with ChainsIndex metadata. My initial fix would simply swap the block of the main block chain with the block being traversed over, but this is incorrect: if C is a fork of B, which is a fork of A, then when blocks from B are moved to A, any preexisting block in A must be moved to C, **not** B. The new invariant caught that mistake.

Also added a test with real world fork data, which triggered the original bug.

As suggested by Greg, includes this **breaking change**:
- renamed RichBlockHeader.chain_ref -> chain_id

### CODE REVIEW

File: crates/btc-relay/src/lib.rs [hunk 1]
```diff
@@ -1094,21 +1094,11 @@ impl<T: Config> Pallet<T> {
...
-                        Self::transfer_block_to_chain(block, MAIN_CHAIN_ID, fork.chain_id)?;
-                    }
-                }
             }
 
-            // transfer the child block to the main chain
-            Self::transfer_block_to_chain(child, child.chain_ref, MAIN_CHAIN_ID)?;
+            // transfer the child block to the main chain, and if main already has a block at this
+            // height, transfer it to `fork`
+            Self::swap_block_to_mainchain(child, child.chain_ref, fork.chain_id)?;
```
Comment (gregdhill) [inline crates/btc-relay/src/lib.rs:1101]:
```suggestion
            Self::swap_block_to_mainchain(child, fork.chain_id)?;
```

File: crates/btc-relay/src/lib.rs [hunk 2]
```diff
@@ -1149,18 +1139,32 @@ impl<T: Config> Pallet<T> {
...
-        new_fork_id: u32,
+        chain_id_for_old_main_blocks: u32,
     ) -> Result<(), DispatchError> {
         let block_height = block.block_height;
+
+        // store the hash we will overwrite, if it exists
+        let replaced_block_hash = ChainsHashes::<T>::try_get(MAIN_CHAIN_ID, block_height);
+
         // remove from old chain and insert into the new one
         ChainsHashes::<T>::remove(old_chain_id, block_height);
```
Comment (gregdhill) [inline crates/btc-relay/src/lib.rs:1155]:
```suggestion
        ChainsHashes::<T>::remove(block.chain_ref, block_height);
```

File: crates/btc-relay/src/lib.rs (hunk 1 above)
Comment (gregdhill) [reply crates/btc-relay/src/lib.rs:1101]: Actually it's not clear to me why this is even called `chain_ref`, can this be renamed to `chain_id`?

Comment (sander2) [reply crates/btc-relay/src/lib.rs]: Good point, renamed.

**Comment by sander2** [comment]
Renamed chain_ref as you suggested. It's a breaking change, though, so I marked it as such. Will add to the breaking changes board

**Review by gregdhill — APPROVED** [review]

**Pull request merged**